### ISSUE TITLE
Add Startup Events - IDnnStartupConfigure

### ISSUE THREAD

**Issue by ahoefling** [dnnsoftware/Dnn.Platform#4167]
## Description of problem
`IDnnStartup` interface is designed to mimic the behaviors of the `Startup.cs` class from AspNetCore. It currently only provides APIs for registering services with the container. We should add a `Configure` method that runs during application startup. 

This will enable DNN extensions to access the DNN Application's startup code and have access to Dependency Injection all at the same time.

## Description of solution
Add new interface `IDnnStartupConfigure` that will define the following contract or similar
```c#
public class IDnnStartupConfigure
{
    void Configure(IApplicationBuilder app, IWebHostEnvironment env);
}
```

## Description of alternatives considered
Not applicable

## Screenshots
Not applicable

## Additional context
Not applicable

## Affected browser
Not applicable

**Comment by bdukes** [comment]
What sorts of things do you expect a developer would be able to do with an `IApplicationBuilder` instance?

I like this idea as a way to run startup logic after the services container is built.  What are some scenarios you see this fulfilling?

**Comment by ahoefling** [comment]
The idea here isn't so much what we can do today, but what we will be able to do tomorrow. Having the plumbing in place will go a very long way. For example Dependency Injection has been in the platform since 9.4 but is really starting to shine with all the enhancements we have made.

Some things I would like to be able to do with `IApplicationBuilder`
- Complete access to the DI container, this allows code to interact with any singletons. The reason why I am proposing this is so we can add an `AggregateEventLogger`. This would enable developers to implement their own `IEventLogger` implementation and when any API is invoked it will write to the DNN event log and wherever the other implementations go. I see an `AggregateLogger` in our future for general logging that we have been talking about
- Adding middleware support where applicable (this is a tricky one)
- This could allow a module or extension to swap out the Dependency Injection container
- Configure a DbContext for entity framework
- Add swashbuckle to some or all Web API routes

This list can go on, but these are some concepts that come to mind

This API structure is using the `Startup.cs` API from AspNetCore 2.1.1 (which makes what is used in DNN today). I think it makes the most sense to include the necessary abstration projects to try and implement this. It will get us 1-step closer to being ready for a .NET Core migration in the future. 

### Implementation Notes
I think a MVP implementation would give us the most basic functionality with these interfaces. Many of the properties on them would throw a `NotSupportedException`. The main goal here is to have a startup event that is triggered during the Application Start sequence after the container is fully built.

### Intent of Issue
We need some type of Startup event that developers can hook into. I am hearing way too many stories of developers adding "Special" startup classes and invoking startup logic in the `RegisterServices` method. The goal of this issue is to bridge that gap so developers can access a startup event that is executed after the container is built.

**Comment by valadas** [comment]
I would love that. One scenario for me was for an extension I had to built that had no entry point was DI was supported and I had to attach to the skinload event. I don't remember how I got around it, but there are certainly some good scenarios.

**Comment by mitchelsellers** [comment]
Yes, i think that this is very valuable, but for sure a more advanced extension point

**Comment by ahoefling** [comment]
I'll submit a draft PR that we can evaluate direction on this. We have 2 different ways to solve this problem that will have either major implications to the platform or minor

## Option 1: Include AspNetCore 2.1 libraries
AspNetCore 2.1 was the last LTS version of .NET Core that was built targeting .NET Framework. If we want to have proper 3rd party tool interoperability we will need to include the correct library. This adds quite a large footprint to DNN with adding 10+ new dlls to the bin directory as well as web.config changes.

## Option 2: Mimic AspNetCore 3.1 or .NET 5 libraries
We can analyze the current .NET 5 startup file specification and make our own `IDnnApplicationBuilder` and other interfaces. These in the future could inherit from .NET 5+ libraries when DNN gets there. This approach would create a minimal footprint but there would be 0 tool interoperability and things would need to be customized

**Comment by ahoefling** [comment]
I submitted a PR that demonstrates how this could look by including AspNetCore 2.1.1 but @mitchelsellers commented on the PR that it wouldn't be accepted (so we need to discuss a new approach)
@mitchelsellers https://github.com/dnnsoftware/Dnn.Platform/pull/4174#issuecomment-706800607
> I believe the inclusion of .NET Core 2.1 assemblies will be a non-starter for us. As we have no upgrade path to get to later versions.

## Proposal
I think it still makes sense to follow the `Startup.cs` paradigm that is popular in both .NET Framework and .NET Core. We would just need to create our own interfaces to inject. 
- IApplicationBuilder - We could create our own carbon copy of this interface called `IDnnApplicationBuilder`. If we get to a point where can reference AspNetCore our interface can just inherit from AspNetCore. This will give us a safe way to move forward. 
- IWebHostEnvironment - I don't see any need to create our own version of this interface at the moment. We already have the `IApplicationInfo` which has some similar properties. If we decide to in the future it can inherit from `IWebHostEnvironment`

### Application Builder
```c#
public interface IDnnApplicationBuilder
{
    IServiceProvider ApplicationServices { get; set; }
}
```

The proposed `IDnnApplicationBuilder` just includes the `IServiceProvider` and nothing else by design. That is because I don't see the need to implement the current features until we have a better understanding of what we want and what we don't want. 

### Configure API
```c#
public interface IDnnStartupWithConfigure
{
    void Configure(IDnnApplicationBuilder applicationBuilder, IApplicationInfo application);
}
```

This contract will give modern .NET developers a similar looking API compared to what they use outside of DNN. It will allow them to perform any application startup logic with a fully built `IServiceProvider`.

### Execution State
In the closed PR #4174 I prototyped how the lifecycle of a `Configure()` API would look. In that example it is invoked at the end of the `Application_Startup()` method which is where I think this belongs. It is essentially being invoked at the very end of the startup routine but will be invoked with a fully built `IServiceProvider`.

## Other Considerations
The goal of this issue was to bridge the gap with our current implementation of `IDnnStartup` which focuses on registering dependencies. This follows established patterns in the .NET community but it can still be confusing.

Do we want to look at creating a simply interface with a focus on lifecycle events?
```public interface IDnnLifecycle
{
    // Invoked during the application startup
    void Start();

    // Invoked when the application is exiting
    void Exit();

    // Invoked if a user navigates to a new tab
    void Sleep();
}
```

This interface would allow developers to implement it and it would automatically inject any service requested
```public class MyDnnLifecycle : IDnnLifecycle
{
    readonly IApplicationInfo applicationInfo;
    public MyDnnLifecycle(IApplicationInfo applicationInfo)
    {
        this.applicationInfo = applicationInfo;
    }

    public void Start() {}
    public void Exit() {}
    public void Sleep() {}
}
```

This style of lifecycle events is very popular in app development paradigms such as mobile and desktop. I don't know if they are useful in DNN which really just cares about an Application Startup event. **I thought it would be useful to look at it for comparison**
